Fix collection data for trashed items

johnswanson opened the pull request:
When something is in the trash, we need to check permissions on the `trashed_from_collection_id` rather than the `collection_id`. We were doing this. However, we want the actual collection data on the search result to represent the actual collection it's in (the trash). I added the code to do this, a test to make sure it works as intended, and a comment to explain what we're doing here and why.

johnswanson committed:
Fix collection IDs for trashed items

When something is in the trash, we need to check permissions on the
`trashed_from_collection_id` rather than the `collection_id`. We were
doing this. However, we want the actual collection data on the search
result to represent the actual collection it's in (the trash). I added
the code to do this, a test to make sure it works as intended, and a
comment to explain what we're doing here and why.

johnswanson commented:
Note that the still-failing e2e tests will be fixed by the UI merging.

noahmoss commented on src/metabase/api/search.clj:
```diff
@@ -106,25 +106,46 @@
 
+(defn- collection-permission-id [model]
+  (case model
+    ("dashboard" "card" "dataset" "action") [:coalesce :trashed_from_collection_id :collection_id]
```
Maybe this list should be defined somewhere closer to the main trash logic? (is that in `metabase.models.collection`?) Just so it's more obvious to update it in case anyone adds `trashed_from_collection_id` to another model in the future

johnswanson committed:
Refactor permission for trashed_from_collection_id

Noah pointed out that the logic of "what collection do I check for
permissions" was getting sprinkled into numerous places, and it felt a
little scary. In fact, there was a bug in the previous implementation.
If you selected a collection with `(t2/select [:model/Collection ...])`,
selecting a subset of columns, and *didn't* include the
`trashed_from_collection_id` in that set of columns, then called
`mi/can-write?` on the result, you'd get erroneous results.
Specifically, we'd return `nil` (representing the root collection).

I think this is a reasonable fix, though I'm pretty new to using fancy
multimethods with `derive` and such. But basically, I wanted a way to
annotate a model to say "check these permissions using
`:trashed_from_collection_id` if the item is archived." And in this
case, we'll throw an exception if `:trashed_from_collection_id` is not
present, just like we currently throw an exception if `:collection_id`
is not present when checking permissions the normal way.

sloansparger commented:
<img width="1491" alt="Screenshot 2024-05-07 at 1 25 04 PM" src="https://github.com/metabase/metabase/assets/7104357/03378197-3d4d-4fd5-a13d-262b4a7dbc1f">
I'm still getting one question that comes back with "Our analytics" as the parent

sloansparger commented:
i'm also seeing it when i view the specific item, so this might not be a search related bug.

<img width="1503" alt="Screenshot 2024-05-07 at 1 26 55 PM" src="https://github.com/metabase/metabase/assets/7104357/b7025831-09bb-4b45-a4f1-2887715f6097">

johnswanson committed:
Rebase fixes
